specifyweb/frontend/js_src/lib/components/Interactions/InteractionDialog.tsx

```diff
@@ -136,39 +149,40 @@ export function InteractionDialog({
     recordSet: SerializedResource<RecordSet> | undefined,
     prepsData: RA<PreparationRow>
   ) {
-    // This is a really ugly piece of code:
     let missing: WritableArray<string> = [];
-    if (Array.isArray(entries)) {
-      let index = 0;
-      let offsetIndex = 0;
-      while (offsetIndex < entries.length && index < prepsData.length) {
-        if (entries[offsetIndex] == prepsData[index][0]) {
-          const value = prepsData[index][0];
-          while (++index < prepsData.length && prepsData[index][0] == value);
-        } else {
-          missing.push(entries[offsetIndex]);
-        }
-        offsetIndex += 1;
-      }
-      if (offsetIndex < entries.length)
-        missing = [...missing, ...entries.slice(offsetIndex)];
+
+    let catalogNumbers: WritableArray<string> = Array.from(
```

Comment:
This code is looking a lot better than what was better but can be improved even further

think about how you could refactor it to not have to use `let` or `WritableArray` at all